clang/test/SemaSYCL/built-in-type-kernel-arg.cpp

```diff
@@ -162,3 +180,15 @@ int main() {
 // CHECK-NEXT: ImplicitCastExpr {{.*}} 'int *' <AddressSpaceConversion>
 // CHECK-NEXT: ImplicitCastExpr {{.*}} '__global int *' <LValueToRValue>
 // CHECK-NEXT: DeclRefExpr {{.*}} '__global int *' lvalue ParmVar {{.*}} '_arg_ptr_array' '__global int *'
+
+// CHECK: FunctionDecl {{.*}}kernel_nns 'void (__generated_test_struct_simple)'
+// CHECK-NEXT: ParmVarDecl {{.*}} used _arg_tds '__generated_test_struct_simple'
+
+// CHECK: VarDecl {{.*}} used __SYCLKernel
+// CHECK: InitListExpr
+// CHECK: CXXConstructExpr {{.*}} 'Nested::TDS':'test_struct_simple' 'void (const test_struct_simple &) noexcept'
+// CHECK: ImplicitCastExpr {{.*}} 'const test_struct_simple' lvalue <NoOp>
+// CHECK: UnaryOperator {{.*}} 'Nested::TDS':'test_struct_simple' lvalue prefix '*' cannot overflow
+// CHECK: CXXReinterpretCastExpr {{.*}} 'Nested::TDS *' reinterpret_cast<struct Nested::TDS *> <BitCast>
```

Comment:
I see the __generated_Type and type we cast to is 'different' because of the typedef. Should we be using canonical type like you did in your offline testing when creating the cast? We should check how clang AST is normally build for this in Godbolt

Reply:
I don't think this needs to hold up this patch though since this is a fix for a crash and is an improvement by itself

Reply:
I see the same thing in Godbolt - https://godbolt.org/z/o1ohPWa1G .

```
`-CXXReinterpretCastExpr <col:57, col:86> 'TD *' reinterpret_cast<TD *> <BitCast>
   `-UnaryOperator <col:79, col:80> 'Nested2 *' prefix '&' cannot overflow
      `-DeclRefExpr <col:80> 'Nested2':'Nested2' lvalue ParmVar 0x563b31016900 'Param2' 'Nested2':'Nested2'

```